Use enum instead of booleans for indicating state

bonomat opened the pull request:
Note: with this we are losing the capability to see if a CFD is in a _weird_ state.

Resolves #2022

bonomat committed:
Use enum instead of booleans for indicating state

Note: with this we are losing the capability to see if a CFD is in a _weird_ state.

Resolves #2022

luckysori commented on daemon/src/position_metrics.rs:
```diff
@@ -158,31 +161,28 @@ impl Cfd {
                 // We give `OfferRejected` priority over `ContractSetupFailed`.
-                if self.is_rejected {
-                    Self { ..self }
+                if let AggregatedState::Rejected = self.state {
```
I would just set it to `Failed`. I don't think we gain anything by reasoning about the previous state.

bonomat replied:
Isn't this the bug what Daniel described?

luckysori replied:
Fair enough: wrongly reporting failed can be problematic because we might think there's a problem when looking at the metrics.

On the other hand, this only applies for CFDs in the `cfds` table. As soon as we move CFDs to the `failed_cfds` table, this problem goes away. That combined with the fact that the comment says that the bug has already been fixed makes me think that we don't need any of this code (or the comment).

bonomat replied:
I'll remove it in a separate commit.

luckysori commented on daemon/src/position_metrics.rs:
```diff
@@ -211,52 +210,44 @@ impl Cfd {
             RefundConfirmed => Self {
-                is_open: false,
-                is_refunded: true,
+                state: AggregatedState::Closed,
```
🔧 

```suggestion
                state: AggregatedState::Refunded,
```

luckysori commented on daemon/src/position_metrics.rs:
```diff
@@ -211,52 +210,44 @@ impl Cfd {
                 // before lock is confirmed. In such a case the CFD is closed
-                is_open: false,
-                is_closed: true,
+                state: AggregatedState::Closed,
```
🤔 It seems weird to overwrite anything here. I would rely on whatever happened before and not do anything here.

bonomat replied:
True, we can just remain in whatever state we have been.

luckysori commented on daemon/src/position_metrics.rs:
```diff
@@ -211,52 +210,44 @@ impl Cfd {
             | CetTimelockExpiredPostOracleAttestation { .. } => Self {
-                is_open: false,
-                is_closed: true,
+                state: AggregatedState::Closed,
```
🤔 Should we even set the `state` to "final" for anything other than `Confirmed` events, rejected and failed? Do we gain anything from reporting this stuff early in the metrics?

bonomat replied:
We would see that a CFD is basically closed at the time the CET expired, not only when the closed transaction was confirmed.

luckysori replied:
Is that preferable when looking at metrics? I just don't know.

bonomat replied:
Yes, because we act on the same information in services built on top and would see discrepancies here.

luckysori reviewed (commented):
Some more suggestions and a little fix.

bonomat committed:
Transition to failed if receiving failed event

If we ever receive a failed event consider this position as failed what state it has been before.

luckysori reviewed (approved):
Oops, forgot to give approval.
